contracts/blender/src/merkle_tree.rs

```diff
@@ -0,0 +1,88 @@
+use ink_prelude::{boxed::Box, vec, vec::Vec};
+use ink_storage::traits::{PackedAllocate, PackedLayout, SpreadAllocate, SpreadLayout};
+use scale::{Decode, Encode};
+#[cfg(feature = "std")]
+use scale_info::TypeInfo;
+
+/// Abstraction for the elements kept in nodes (both in leaves and inner nodes).
+#[cfg(feature = "std")]
+pub trait TreeElement: Clone + PackedLayout + PackedAllocate + TypeInfo + 'static {}
+#[cfg(feature = "std")]
+impl<T: Clone + PackedLayout + PackedAllocate + TypeInfo + 'static> TreeElement for T {}
+
+#[cfg(not(feature = "std"))]
+pub trait TreeElement: Clone + PackedLayout + PackedAllocate {}
+#[cfg(not(feature = "std"))]
+impl<T: Clone + PackedLayout + PackedAllocate> TreeElement for T {}
+
+/// Abstraction for a two-to-one hashing function.
+///
+/// This is used to compute hash in parent node from the children ones.
+pub trait KinderBlender<H>: PackedLayout + Default {
+    /// Compute hash from two hashes.
+    fn blend_kinder(left: &H, right: &H) -> H;
+}
+
+/// Simplified binary tree that represents a Merkle tree over some set of hashes.
+///
+/// It has `LEAVES` leaves (and thus `2 * LEAVES - 1` nodes in general). `KB` is used for computing
+/// values in the inner nodes. `TE` represents *both* leaves' values and inner nodes' values.
+///
+/// `LEAVES` must be power of `2`.
+#[derive(Clone, Eq, PartialEq, Decode, Encode, PackedLayout, SpreadLayout, SpreadAllocate)]
+#[cfg_attr(feature = "std", derive(TypeInfo, ink_storage::traits::StorageLayout))]
+pub struct MerkleTree<TE: TreeElement, KB: KinderBlender<TE>, const LEAVES: usize> {
+    /// Array of node values (root is at [1], children are in [2n] and [2n+1]).
+    nodes: Vec<TE>,
+    /// Marker of the first 'non-occupied' leaf.
+    next_free_leaf: u32,
+    #[codec(skip)]
+    _phantom: Box<KB>,
+}
+
+/// Creates an 'empty' tree, i.e. every leaf contains `TE::default()`.
+impl<TE: TreeElement + Default, KB: KinderBlender<TE>, const LEAVES: usize> Default
```

Comment:
Why do you need the `KinderBlender` abstraction? Do you really see it being changed?

Reply:
And why not use the same hasher as in real contract? You need to test what you're gonna ship anyway.

Reply:
I was just thinking that this additional complexity may not be worth the gains here but if you disagree I'm fine with that as well.

Reply:
let's wait for other reviewers to settle this one :)

Reply:
To me this seems like an instance of YAGNI.

Reply:
![image](https://user-images.githubusercontent.com/27450471/201681227-81c31cc1-3488-44e8-95b2-92fac1cfa698.png)

Reply:
Ready to resolve when you are 👍